keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -0,0 +1,383 @@
+and I want to allow the egress of the traffic only the NodePort range 
+(eg. 30000-32767) as I don't know which port is going to be allocated on the 
+other side, but don't want to create a rule for each of them.
+
+#### Story 2 - Blocking the egress for not allowed insecure ports
+As a developer, I need to create an application that scrapes informations from 
+multiple sources, being those sources databases running in random ports, web 
+applications and other sources. But the security policy of my company asks me 
+to block communication with well known ports, like 111 and 445, so I need to create
+a network policy that allows me to communicate with any port except those two and so
+I can be compliant with the company's policy.
+
+#### Story 3 - Containerized Passive FTP Server
+As a Kubernetes User, I've received a demand from my boss to run our FTP server in an 
+existing Kubernetes cluster, to support some of my legacy applications. 
+This FTP Server must be acessible from inside the cluster and outside the cluster, 
+but I still need to keep the basic security policies from my company, that demands 
+the existence of a default deny rule for all workloads and allowing only specific ports.
+
+Because this FTP Server runs in PASV mode, I need to open the Network Policy to ports 21 
+and also to the range 49152-65535 without allowing any other ports.
+
+
+### Notes/Constraints/Caveats
+
+*  The technology used by the CNI provider might not support port range in a 
+trivial way as described in [#drawbacks]
+
+### Risks and Mitigations
+
+CNIs will need to support the new field in their controllers. For this case 
+we'll try to make broader communication with the main CNIs so they can be aware
+of the new field.
+
+## Design Details
+
+API changes to NetworkPolicy:
+* Add a new field called ``EndPort`` inside ``NetworkPolicyPort`` as the following:
+```
+// NetworkPolicyPort describes a port to allow traffic on
+type NetworkPolicyPort struct {
+	// The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
+	// field defaults to TCP.
+	// +optional
+	Protocol   *v1.Protocol `json:"protocol,omitempty" protobuf:"bytes,1,opt,name=protocol,casttype=k8s.io/api/core/v1.Protocol"`
+
+	// The port on the given protocol. This can either be a numerical or named port on
+	// a pod. If this field is not provided, this matches all port names and numbers.
+	// +optional
+	Port       *intstr.IntOrString `json:"port,omitempty" protobuf:"bytes,2,opt,name=port"`
+
+	// EndPort defines the end of the port range, being the end included within the
+	// range
+	// +optional
+	EndPort    int32 `json:"port,omitempty" protobuf:"bytes,2,opt,name=endPort"`
+}
+```
+
+### Validations
+The ``NetworkPolicyPort`` will need to be validated, with the following scenarios:
+* If an ``EndPort`` is specified a ``Port`` must also be specified
+* If ``Port`` is a string ``EndPort`` cannot be specified
```

Comment:
I think this limitation is OK, but not strictly required

Reply:
what would be the interpretation of 

```
Port: http
EndPort: 9000
```
?
maybe it is better to take a stand and not allow this?

Reply:
+1 not to allowing this. This can lead to a conversion confusion to CNIs. One concern I had was: does iptables supports named ports? Otherwise we could assume those ports are always converted to int by the CNI before inserting into the dataplane.

The answer is: yes, it accepts. Because we don't know how this is done in the wild (if there's some CNI creating rules with the named ports, and not the numbered ones) I think it's better also to say 'if you want a range, you can only use integer as the start of the range'

Reply:
Whatever implements policy must convert named ports to numeric.

We *could* allow range here but it feels like a super corner case and let's just not.